### ISSUE TITLE
Lack of parallelism between avg_predictions() and avg_comparisons()

### ISSUE THREAD

**Issue by BorgeJorge** [vincentarelbundock/marginaleffects#766]
Forgive me if this is my misunderstanding and not an actual "issue."

Say I want to estimate predicted levels of an outcome at different levels of treatment, as well as average contrasts or average marginal effects, for a full sample and for racial subgroups. Using the Lalonde dataset:

```
data('lalonde', package='MatchIt')
library(marginaleffects)

fit <- lm(re78 ~ race * treat, data = lalonde)
avg_comparisons(fit, variables = "treat")
avg_predictions(fit, variables = "treat")
avg_comparisons(fit, variables = "treat", by = 'race')
avg_predictions(fit, variables = "treat", by = 'race')
```

The first call to avg_comparisons() returns the average contrast between predicted values when treat = 1 vs. when treat = 0 for the full sample, and the first call to avg_predictions returns the average predicted levels for when treat = 1 and for when treat = 0 for the full sample. In other words, the avg_predictions() result has twice as many rows as the avg_comparisons() result.

The second call to avg_comparisons() returns an average treatment contrast for each racial subgroup, as I'd expect. But the second call to avg_predictions() no longer returns twice the number of rows as its corresponding avg_comparisons() call. It just gives overall average predictions for each racial subgroup, and not for each treatment level within each racial subgroup, which latter is what I would expect.

Is there another way to get predicted average outcomes for each treatment level within each racial subgroup?

One other question: The average treatment contrast for the full sample from the first avg_comparisons() call is 560. The two average predictions from the first avg_predictions() call are 6993 (for treat=1) and 6525 (for treat=0). But 6993 - 6525 = 468, not 560. When I convert the data to Stata and run the following in Stata:

```
reg re78 race##treat
margins treat
margins r.treat
```

I get the same average contrast (560) but different average predicted levels: 7024 (for treat=1) and 6464 (for treat=0), whose difference is exactly 560. Are there other commands I should be using if I want to replicate Stata's margins results using marginaleffects?

Thanks so much!
Tod Mijanovich

**Comment by vincentarelbundock** [comment]
You probably want: 

```r
avg_predictions(fit, variables = "treat", by = c('treat', 'race'))
```

Can you please read the Stata comparison vignette and let me know if you still have questions after?

https://vincentarelbundock.github.io/marginaleffects/articles/alternative_software.html#stata

**Comment by ngreifer** [comment]
@BorgeJorge `avg_predictions()` with `variables` supplied and `by` omitted automatically supplied the variable named in `variables` to `by` as a shortcut to avoid having to type it twice. But when you supply your own argument to `by`, you are overriding that functionality and therefore need to specify `by` fully, i.e., by including the variable named in `variables` and the subgrouping variable. Vincent's solution does exactly this.

@vincentarelbundock, maybe it makes sense to add the variable named in `variables` to `by` automatically even if the user supplies their own `by`. I cannot imagine a single scenario in which you would not want averaged predictions split by the counterfactual-inducing variable (i.e., named in `variables`) in addition to whatever variable you would want to subgroup by. That is, I agree with OP's confusion and agree that the two should be consistent.

**Comment by vincentarelbundock** [comment]
Hmmm, ok. 

What's the interface with someone does NOT want to do this automatically? (I can't think of a use case right now either, but I've found that people come up with uses cases that I'm surprised by, so I'm wary for unreversible automagical solutions)

**Comment by BorgeJorge** [comment]
@vincentarelbundock and @ngreifer, thanks very much for clarifying what's going on with `variables` and `by`, it now makes perfect sense!

May I ask, though, about the discrepancy between the AME returned by `avg_comparisons()` and the difference between the two predictions returned by `avg_predictions()` after estimating a model with interactions? To give the example from my first post, when I run the following:

```
data('lalonde', package='MatchIt')
library(marginaleffects)

fit <- lm(re78 ~ race * treat, data = lalonde)
avg_comparisons(fit, variables = "treat")
avg_predictions(fit, variables = "treat")
```

The contrast is 560, but the difference between the two predictions is 468. Shouldn't these be equal? Indeed, when you regress `re78` on `treat` alone, the contrast is exactly equal to the difference between the predictions, so something about the interaction introduces the discrepancy.

If it helps, the following code actually *does* produce a contrast of 468 rather than 560, for each level of treat:

```
fit <- lm(re78 ~ race * treat, data = lalonde)
avg_comparisons(fit, by = "treat")
```

Interestingly enough, the above also produces racial contrasts, again for each level of treat.

@vincentarelbundock, thanks for pointing me to the Stata comparison vignette. I assume you wanted me to see Stata's dydx syntax? The dydx syntax returns results that are equivalent to the syntax I used (`margins treat` and `margins r.treat`), which latter is Stata's syntax for returning contrasts and predictions when using factor variables. Stata always notes this equivalence at the bottom of dydx results, with the following:

```
Note: dy/dx for factor levels is the discrete change from the base level.
```

**Comment by vincentarelbundock** [comment]
@BorgeJorge,

Thanks a lot for this last question and for your patience. I was finally able to look into this more carefully, and it looks like you discovered a bug which occurred in cases when:

1. `avg_predictions()` is used (*not* `predictions()`)
2. All predictors are binary, character, or factor
3. The `wts` argument is not used.
4. The `variables` argument is used.

In principle, this should be fixed in version 0.11.1.9012 on Github.

I added various tests based on your example here: https://github.com/vincentarelbundock/marginaleffects/blob/main/inst/tinytest/test-datagrid.R#L78

Technical background for reference: In the cases described above, we take unique combinations of rows and weights instead of making predictions for all rows of the original data. This can be much faster. Unfortunately, the `datagridcf()` call which is triggered by the `variables` argument dropped the `marginaleffects_wts_internal` column.

**Comment by BorgeJorge** [comment]
Thanks so much for letting me know, @vincentarelbundock, and for addressing this so quickly! I'm really appreciative of all the work you've put into what's become an essential package for my work.
